[ADD] estate: Add initial module structure with property model

SurryaT10 opened the pull request:
Initialized Estate module and created 'estate_property' model.

Commit:
[ADD] estate: Add initial module structure with property model and manifest.

Created 'estate_property' model and fields.

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1072.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1072)

lost-odoo commented:
Hello, can you please ensure that your runbot is green :smile:

Commit:
[FIX] Fixed CI test issues

Commit:
[Add] Added Estate menu and Properties action for estate.property model

Also added access rights for estate.property model

Commit:
[FIX] Added missed imports for views and security in manifest file

lost-odoo commented on estate/models/estate.py:
```diff
@@ -0,0 +1,22 @@
+    _name = 'estate.property'
+    _description = 'Estate Property'
+
+    name = fields.Char('Property Name', required=True)
```
We use double quotes for strings that will be shown on the view. So instead of `'Property Name'` it is better to do `"Property Name"`.

lost-odoo commented on estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,8 @@
+            <menuitem id="estate_menu_properties" name="Properties" action="estate_property_action"/>
+        </menuitem>
+    </menuitem>
+</odoo>
```
Don't forget to always add a final new line :smile:

lost-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,31 @@
+    'author': "Odoo",
+    'website': "https://www.odoo.com",
+
+    # Categories can be used to filter modules in modules listing
```
You can remove those comments here and under from the manifest file.

lost-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,31 @@
+        'security/ir.model.access.csv',
+    ],
+    'application': True,
+    'installable': True,
```
By default when `application` is set, `installable` is set to true automatically.

lost-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,31 @@
+    'name': "Estate",
+
+    'summary': """
+        Starting module for "Estate"
```
I don't think `Starting module for "Estate"` in the summary and description are really appropriate here.

lost-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,8 @@
+        <field name="res_model">estate.property</field>
+        <field name="view_mode">list,form</field>
+    </record>
+</odoo>
```
Here too

lost-odoo reviewed (changes requested):
Hello, I already made a small review. Not a big deal but just some nitpicking stuff that we try to avoid. :smile:

lost-odoo commented:
Also please don't forget to make one commit per chapter. Try to make a correct title. Don't hesitate to amend or squash your commits.

Commit:
[ADD] Chapter 6: Build Views

Added estate_property_views.xml which contains list and form views

Added a separate estate_property_search_view.xml for search with filters for active and availability as well as group by based on postcode.

Also addressed code review comments.

Commit:
[ADD] Chapter 7: Relations Between Models

Added estate.property.type model which specifies the type of property. Eg: House/Apartment

Added estate.property.tag model to indicate the tags(cozy, renovated) for each property

Added estate.property.offer model which indicates the offers for each property.

Also added views for each model mentioned above.

lost-odoo commented:
Hey don't forget to close all the comments I did on the review when you did them :smile: 
Also please follow the [git guidelines](https://www.odoo.com/documentation/19.0/contributing/development/git_guidelines.html). ADD should only be used when you add a new module otherwise use IMP for an improvement.

lost-odoo committed:
DO NOT REMOVE THIS COMMIT

Commit:
[IMP] Added computed fields - total_area and best_offer

total_area field is based on the size of both garden_area and living_area

best_offer field is based on the max price among the list of offers

Also added validity and deadline for the offers as a computed and inverse function impacting each other.

Commit:
[IMP] Chapter 9: Ready for Some Action

Added Sold and Cancel buttons:
When a property is sold it cannot be canceled
When a property is canceled it cannot be sold

Added offer accept and rejected icon button:
When an offer is accepted:
 1. The selling price is automatically updated by button actions
 2. Status changes to 'Offer Accepted'
 3. Rejects all other offers

 Added  an onchange, which updates the status to 'Offer Received' when there is an offer.

Commit:
[IMP] Chapter 10: Constraints

Added SQL constraints for expected_price, selling_price, property_tag_name and property_type_name

Added python constraint for selling_price to be >= 90% of expected_price unless it is 0.0

lost-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,22 @@
+    'website': "https://www.odoo.com",
+
+    'category': 'Tutorials',
+    'version': '0.1',
```
You can switch this to 1.0 which is the default for a new module.

lost-odoo commented on estate/models/estate.py:
```diff
@@ -0,0 +1,94 @@
+        required=True)
+    status = fields.Selection(
+        string="Status",
+        selection=[('new', "New"), ('offer_received', "Offer Received"), ('offer_accepted', "Offer Accepted"), ('sold', "Sold"), ('canceled', "Canceled")],
```
Can you split this into multiple lines please. Same for above.

lost-odoo commented on estate/models/estate.py:
```diff
@@ -0,0 +1,94 @@
+            if record.status == 'new' and len(record.offer_ids) > 0:
+                record.status = 'offer_received'
+            else:
+                record.status = record.status
```
This is not needed anymore as you changed the compute to onchange.

lost-odoo commented on estate/models/estate.py:
```diff
@@ -0,0 +1,94 @@
+    def action_mark_sold(self):
+        for record in self:
+            if record.status == 'canceled':
+                raise UserError("Canceled properties cannot be sold.")
```
You need to translate this string. You can use `self.env._("Canceled properties cannot be sold.")` to make it translatable.

lost-odoo commented on estate/models/estate.py:
```diff
@@ -0,0 +1,94 @@
+    def action_mark_cancel(self):
+        for record in self:
+            if record.status == 'sold':
+                raise UserError("Sold properties cannot be canceled.")
```
Same here

lost-odoo commented on estate/models/estate.py:
```diff
@@ -0,0 +1,94 @@
+    _check_selling_price = models.Constraint(
+        'CHECK(selling_price >= 0)',
+        'The selling price must be positive'
+    )
```
Can you move that on top of the class. It seems a bit lost here, I would put it after or before the fields but, before any functions.

lost-odoo commented on estate/models/estate.py:
```diff
@@ -0,0 +1,94 @@
+    def _check_selling_price(self):
+        for record in self:
+            if not float_is_zero(record.selling_price, precision_digits=2):
+                if float_compare(record.selling_price, record.expected_price * 0.9, precision_digits=2) < 0:
```
Why two conditions ? I think it could be merged into one.

lost-odoo commented on estate/models/estate.py:
```diff
@@ -0,0 +1,94 @@
+        for record in self:
+            if not float_is_zero(record.selling_price, precision_digits=2):
+                if float_compare(record.selling_price, record.expected_price * 0.9, precision_digits=2) < 0:
+                    raise ValidationError('The selling price must be atleast 90% of the expected price! You must reduce the expected price to accept the offer')
```
Don't forget about the double quotes here :smile:

lost-odoo commented on estate/models/property_tag.py:
```diff
@@ -0,0 +1,13 @@
+
+class PropertyTag(models.Model):
+    _name = 'estate.property.tag'
+    _description = 'Estate Property Tag'
```
Here this can be double quotes

lost-odoo commented on estate/models/offer.py:
```diff
@@ -0,0 +1,51 @@
+
+    _check_price = models.Constraint(
+        'CHECK(price > 0)',
+        'The offer price must be strictly positive'
```
Here this can also be double quotes.

Commit:
[IMP] Chapter 11: Add The Sprinkles

Added ordering for the following:
property: Descending by ID
offer: Descending by price
tag: Name
type: Name

Added sequence ordering for type

Added color field for property tag.

Added invisible attribute for Sold, Cancel, Accept, Reject buttons

Added color decorations for list view.

Added offer count button to show the list of offers for a property type

Fixed code review changes.

Commit:
[IMP] Chapter12: Inheritance

Added property_ids field for res.users model

Added the 'property_ids' field in 'Properties for Sale' page in 'Users & Companies' menu

Also removed trailing white spaces in estate.py selection fields

Commit:
[IMP] Chapter12: Inheritance

Added property_ids field for res.users model

Added the 'property_ids' field in 'Properties for Sale' page in 'Users & Companies' menu

Also removed trailing white spaces in estate.py selection fields

Commit:
[ADD] Chapter 13: Interact With Other Modules

Added estate_account module

Overrided the action_mark_sold method to create an invoice when property is marked as Sold

Added 2 invoice lines:
one for 6% of selling price - brokerage fee
one for 100 - administrative fee

Commit:
[IMP] Chapter 14: A Brief History of QWeb

Added a Kandban view for Properties

LoganStaelens commented on estate/models/offer.py:
```diff
@@ -0,0 +1,65 @@
+    @api.model
+    def create(self, vals_list):
+        for record in vals_list:
+            property = self.env['estate.property'].browse(record["property_id"])
```
Avoid browsing in a loop, try to batch it by making the browse outside the loop

LoganStaelens commented on estate/models/estate.py:
```diff
@@ -0,0 +1,108 @@
+
+    def action_mark_sold(self):
+        for record in self:
+            if record.status == "canceled":
```
simple quotes :smile: 
Same at other places

LoganStaelens commented on estate/models/estate.py:
```diff
@@ -0,0 +1,108 @@
+            record.best_price = max(record.offer_ids.mapped('price')) if record.offer_ids else 0.0
+
+    @api.onchange('offer_ids')
+    def _onchange_property_status(self):
```
I think it should be named `_onchange_offer_ids` because it is offer_ids that is changing :smile:

LoganStaelens commented on estate/models/offer.py:
```diff
@@ -0,0 +1,65 @@
+    def action_reject_offer(self):
+        for record in self:
+            record.status = "refused"
+            if record.property_id.status == "offer_accepted" and record.property_id.buyer_id == record.partner_id:
+                record.property_id.selling_price = 0.0
+                record.property_id.buyer_id = False
+                record.property_id.status = "offer_received"
```
Does this make sense to reject an offer that got accepted ? Can a buyer make multiple offers ? if so then it is wrong I guess ?

LoganStaelens commented on estate/models/offer.py:
```diff
@@ -0,0 +1,65 @@
+    def create(self, vals_list):
+        for record in vals_list:
+            property = self.env['estate.property'].browse(record["property_id"])
+            min_price = min(offer.price for offer in property.offer_ids) if property.offer_ids else 0
```
I don't think you are supposed to do that. The minimum price is 90 % of the property asked price.

Also when you do float comparisons, use float_compare.

LoganStaelens commented on estate/models/offer.py:
```diff
@@ -0,0 +1,65 @@
+    @api.depends('validity', 'create_date')
+    def _compute_date_deadline(self):
+        for record in self:
+            create_date = record.create_date if record.create_date else fields.Date.today()
```
```suggestion
            create_date = record.create_date or fields.Date.today()
```

LoganStaelens commented on estate/models/offer.py:
```diff
@@ -0,0 +1,65 @@
+    def _compute_date_deadline(self):
+        for record in self:
+            create_date = record.create_date if record.create_date else fields.Date.today()
+            record.date_deadline = fields.Date.add(create_date, days=record.validity)
```
you could use relativedelta :smile: But it is the same output.

LoganStaelens commented on estate/models/property_tag.py:
```diff
@@ -0,0 +1,15 @@
+    color = fields.Integer("Color")
+
+    _unique_name = models.Constraint(
+        "UNIQUE(name)",
```
single quotes :eyes:

LoganStaelens commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,122 @@
+                <templates>
+                    <t t-name="card">
+                        <field name="name" class="fw-bold"/>
+                        <field name="status" invisible="1"/>
```
You could move that out of the card template and remove the invisible="1".

LoganStaelens commented on estate_account/__manifest__.py:
```diff
@@ -0,0 +1,14 @@
+    'depends': ['base', 'web', 'estate', 'account'],
+    'data': [],
+    'application': True,
+    'license': 'AGPL-3'
```
Do not forget auto_install to auto install this module when all its dependencies are installed.

LoganStaelens commented on estate_account/__manifest__.py:
```diff
@@ -0,0 +1,14 @@
+    'category': 'Tutorials',
+    'version': '1.0',
+
+    'depends': ['base', 'web', 'estate', 'account'],
```
Base and web are useless here as estate and account already depends on them :smile:

LoganStaelens commented on estate_account/models/estate_property.py:
```diff
@@ -0,0 +1,26 @@
+            self.env['account.move'].create({
+                'partner_id': record.buyer_id.id,
+                'move_type': "out_invoice",
+                'journal_id': self.env['account.journal'].search([('type', '=', 'sale')], limit=1).id,
```
Try to avoid doing search in loops even if it is unlikely to have multiple records to mark as sold at once :smile:
